File: llm_written/claude-sonnet-4-5/bfs/java/TestBfs.java
import java.util.*;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class TestBfs {
    private static final int NUM_DETERMINISM_RUNS = 3;
    
    public static void main(String[] args) {
        TestBfs tester = new TestBfs();
        boolean allPassed = true;

        System.out.println("=== BFS Differential Testing ===\n");

        // Edge cases
        allPassed &= tester.testEmptyGraph();
        allPassed &= tester.testSingleVertex();
        allPassed &= tester.testDisconnectedGraph();
        allPassed &= tester.testInvalidStartVertex();

        // Small cases
        allPassed &= tester.testLinearChain(10);
        allPassed &= tester.testCompleteBinaryTree(7);
        allPassed &= tester.testCycle(5);

        // Medium cases
        allPassed &= tester.testGrid(10, 10);
        allPassed &= tester.testRandomGraph(50, 100);

        // Large cases
        allPassed &= tester.testGrid(50, 50);
        allPassed &= tester.testRandomGraph(1000, 5000);
        allPassed &= tester.testCompleteBinaryTree(1023);

        System.out.println("\n=== Summary ===");
        if (allPassed) {
            System.out.println("✓ All tests PASSED");
            System.exit(0);
        } else {
            System.out.println("✗ Some tests FAILED");
            System.exit(1);
        }
    }

    private boolean testEmptyGraph() {
        System.out.println("Test: Empty Graph");
        Graph g = new Graph();
        return runTest("EmptyGraph", g, 0);
    }

    private boolean testSingleVertex() {
        System.out.println("Test: Single Vertex");
        Graph g = new Graph();
        g.getVertices().put(0, new ArrayList<>());
        return runTest("SingleVertex", g, 0);
    }

    private boolean testDisconnectedGraph() {
        System.out.println("Test: Disconnected Graph");
        Graph g = new Graph();
        g.addEdge(0, 1);
        g.addEdge(1, 2);
        g.addEdge(3, 4);
        g.addEdge(4, 5);
        return runTest("DisconnectedGraph", g, 0);
    }

    private boolean testInvalidStartVertex() {
        System.out.println("Test: Invalid Start Vertex");
        Graph g = new Graph();
        g.addEdge(0, 1);
        g.addEdge(1, 2);
        return runTest("InvalidStartVertex", g, 99);
    }

    private boolean testLinearChain(int length) {
        System.out.println("Test: Linear Chain (length=" + length + ")");
        Graph g = new Graph();
        for (int i = 0; i < length - 1; i++) {
            g.addEdge(i, i + 1);
        }
        return runTest("LinearChain_" + length, g, 0);
    }

    private boolean testCycle(int size) {
        System.out.println("Test: Cycle (size=" + size + ")");
        Graph g = new Graph();
        for (int i = 0; i < size; i++) {
            g.addEdge(i, (i + 1) % size);
        }
        return runTest("Cycle_" + size, g, 0);
    }

    private boolean testCompleteBinaryTree(int numVertices) {
        System.out.println("Test: Complete Binary Tree (vertices=" + numVertices + ")");
        Graph g = new Graph();
        for (int i = 0; i < numVertices; i++) {
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            if (left < numVertices) {
                g.addEdge(i, left);
            }
            if (right < numVertices) {
                g.addEdge(i, right);
            }
        }
        return runTest("BinaryTree_" + numVertices, g, 0);
    }

    private boolean testGrid(int rows, int cols) {
        System.out.println("Test: Grid (rows=" + rows + ", cols=" + cols + ")");
        Graph g = new Graph();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                int v = r * cols + c;
                if (c < cols - 1) {
                    g.addEdge(v, v + 1);
                }
                if (r < rows - 1) {
                    g.addEdge(v, v + cols);
                }
            }
        }
        return runTest("Grid_" + rows + "x" + cols, g, 0);
    }

    private boolean testRandomGraph(int vertices, int edges) {
        System.out.println("Test: Random Graph (V=" + vertices + ", E=" + edges + ")");
        Graph g = new Graph();
        Random rand = new Random(42); // Fixed seed for reproducibility
        
        // Ensure all vertices exist
        for (int i = 0; i < vertices; i++) {
            g.getVertices().putIfAbsent(i, new ArrayList<>());
        }
        
        // Add random edges
        Set<String> addedEdges = new HashSet<>();
        int edgeCount = 0;
        while (edgeCount < edges) {
            int u = rand.nextInt(vertices);
            int v = rand.nextInt(vertices);
            if (u != v) {
                String edge = Math.min(u, v) + "-" + Math.max(u, v);
                if (!addedEdges.contains(edge)) {
                    g.addEdge(u, v);
                    addedEdges.add(edge);
                    edgeCount++;
                }
            }
        }
        
        return runTest("RandomGraph_V" + vertices + "_E" + edges, g, 0);
    }

    private boolean runTest(String testName, Graph graph, int startVertex) {
        try {
            // Run sequential baseline
            List<Integer> seqResult = BfsSequential.run(graph, startVertex);
            String seqHash = hashList(seqResult);

            // Run parallel version multiple times for determinism check
            BfsParallel bfsParallel = new BfsParallel();
            List<String> parHashes = new ArrayList<>();
            List<Integer> firstParResult = null;

            for (int i = 0; i < NUM_DETERMINISM_RUNS; i++) {
                List<Integer> parResult = bfsParallel.run(graph, startVertex);
                if (i == 0) {
                    firstParResult = parResult;
                }
                parHashes.add(hashList(parResult));
            }
            bfsParallel.shutdown();

            // Check correctness (parallel matches sequential)
            boolean correctness = seqHash.equals(parHashes.get(0));
            
            // Check determinism (all parallel runs match)
            boolean determinism = parHashes.stream().allMatch(h -> h.equals(parHashes.get(0)));

            if (correctness && determinism) {
                System.out.println("  ✓ PASS - Correctness: YES, Determinism: YES");
                System.out.println("    Sequential hash: " + seqHash);
                System.out.println("    Parallel hash:   " + parHashes.get(0));
                return true;
            } else {
                System.out.println("  ✗ FAIL");
                if (!correctness) {
                    System.out.println("    Correctness FAILED:");
                    System.out.println("      Sequential: " + seqResult);
                    System.out.println("      Parallel:   " + firstParResult);
                    System.out.println("      Seq hash: " + seqHash);
                    System.out.println("      Par hash: " + parHashes.get(0));
                }
                if (!determinism) {
                    System.out.println("    Determinism FAILED:");
                    for (int i = 0; i < parHashes.size(); i++) {
                        System.out.println("      Run " + (i + 1) + ": " + parHashes.get(i));
                    }
                }
                return false;
            }
        } catch (Exception e) {
            System.out.println("  ✗ EXCEPTION: " + e.getMessage());
            e.printStackTrace();
            return false;
        }
    }

    private String hashList(List<Integer> list) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            String str = list.toString();
            byte[] hash = md.digest(str.getBytes());
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString().substring(0, 16);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

File: llm_written/claude-sonnet-4-5/bfs/cpp/test_bfs.cpp
#include "bfs_seq.hpp"
#include "bfs_parallel.hpp"
#include "graph.h"
#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <cstring>
#include <algorithm>
#include <omp.h>

// Simple hash function for vector<int>
size_t hash_vector(const std::vector<int>& vec) {
    size_t hash = vec.size();
    for (int v : vec) {
        hash ^= std::hash<int>{}(v) + 0x9e3779b9 + (hash << 6) + (hash >> 2);
    }
    return hash;
}

// Convert hash to hex string
std::string hash_to_hex(size_t hash) {
    std::stringstream ss;
    ss << std::hex << std::setfill('0') << std::setw(16) << hash;
    return ss.str();
}

// Test result structure
struct TestResult {
    std::string name;
    bool correctness_pass;
    bool determinism_pass;
    std::vector<size_t> hashes;
    int size;
    std::string error_msg;
};

// Create a line graph: 0-1-2-3-...-n
Graph create_line_graph(int n) {
    Graph g;
    for (int i = 0; i < n - 1; i++) {
        g.add_edge(i, i + 1);
    }
    return g;
}

// Create a grid graph: n x n grid
Graph create_grid_graph(int n) {
    Graph g;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int v = i * n + j;
            if (j < n - 1) g.add_edge(v, v + 1);  // right
            if (i < n - 1) g.add_edge(v, v + n);  // down
        }
    }
    return g;
}

// Create a star graph: center connected to n spokes
Graph create_star_graph(int n) {
    Graph g;
    for (int i = 1; i <= n; i++) {
        g.add_edge(0, i);
    }
    return g;
}

// Create a complete graph: all vertices connected
Graph create_complete_graph(int n) {
    Graph g;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            g.add_edge(i, j);
        }
    }
    return g;
}

// Run a single test case
TestResult run_test(const std::string& name, Graph& g, int start, int num_runs = 3) {
    TestResult result;
    result.name = name;
    result.size = g.vertices.size();
    result.correctness_pass = false;
    result.determinism_pass = false;
    
    // Run sequential baseline
    std::vector<int> seq_result;
    try {
        seq_result = bfs(g, start);
    } catch (const std::exception& e) {
        result.error_msg = std::string("Sequential failed: ") + e.what();
        return result;
    }
    
    size_t seq_hash = hash_vector(seq_result);
    
    // Run parallel version multiple times
    std::vector<std::vector<int>> par_results;
    for (int run = 0; run < num_runs; run++) {
        try {
            std::vector<int> par_result = bfs_parallel(g, start);
            par_results.push_back(par_result);
            result.hashes.push_back(hash_vector(par_result));
        } catch (const std::exception& e) {
            result.error_msg = std::string("Parallel run ") + std::to_string(run) + " failed: " + e.what();
            return result;
        }
    }
    
    // Check correctness: first parallel run should match sequential
    result.correctness_pass = (par_results[0] == seq_result);
    
    if (!result.correctness_pass) {
        result.error_msg = "Output mismatch. Seq size: " + std::to_string(seq_result.size()) + 
                          ", Par size: " + std::to_string(par_results[0].size());
        return result;
    }
    
    // Check determinism: all parallel runs should produce identical results
    result.determinism_pass = true;
    for (size_t i = 1; i < par_results.size(); i++) {
        if (par_results[i] != par_results[0]) {
            result.determinism_pass = false;
            result.error_msg = "Determinism failure: run 0 vs run " + std::to_string(i);
            break;
        }
    }
    
    return result;
}

// Performance benchmark
struct PerfResult {
    std::string name;
    int size;
    double seq_time_ms;
    double par_time_ms;
    double speedup;
    int num_threads;
};

PerfResult run_perf_test(const std::string& name, Graph& g, int start) {
    PerfResult result;
    result.name = name;
    result.size = g.vertices.size();
    result.num_threads = omp_get_max_threads();
    
    // Warm-up
    bfs(g, start);
    bfs_parallel(g, start);
    
    // Sequential timing (3 runs, take median)
    std::vector<double> seq_times;
    for (int i = 0; i < 3; i++) {
        auto start_time = std::chrono::high_resolution_clock::now();
        bfs(g, start);
        auto end_time = std::chrono::high_resolution_clock::now();
        seq_times.push_back(std::chrono::duration<double, std::milli>(end_time - start_time).count());
    }
    std::sort(seq_times.begin(), seq_times.end());
    result.seq_time_ms = seq_times[1];
    
    // Parallel timing (3 runs, take median)
    std::vector<double> par_times;
    for (int i = 0; i < 3; i++) {
        auto start_time = std::chrono::high_resolution_clock::now();
        bfs_parallel(g, start);
        auto end_time = std::chrono::high_resolution_clock::now();
        par_times.push_back(std::chrono::duration<double, std::milli>(end_time - start_time).count());
    }
    std::sort(par_times.begin(), par_times.end());
    result.par_time_ms = par_times[1];
    
    result.speedup = result.seq_time_ms / result.par_time_ms;
    
    return result;
}

int main(int argc, char* argv[]) {
    bool run_perf = false;
    if (argc > 1 && std::strcmp(argv[1], "--perf") == 0) {
        run_perf = true;
    }
    
    std::vector<TestResult> test_results;
    
    std::cout << "=== BFS Differential Testing ===" << std::endl;
    std::cout << "Running correctness and determinism tests..." << std::endl << std::endl;
    
    // Edge cases
    {
        Graph empty_g;
        test_results.push_back(run_test("Empty graph", empty_g, 0));
    }
    
    {
        Graph single_g;
        single_g.vertices[0] = {};
        test_results.push_back(run_test("Single vertex", single_g, 0));
    }
    
    {
        Graph g = create_line_graph(5);
        test_results.push_back(run_test("Start not in graph", g, 99));
    }
    
    // Small graphs
    {
        Graph g = create_line_graph(10);
        test_results.push_back(run_test("Line graph (10)", g, 0));
    }
    
    {
        Graph g = create_star_graph(20);
        test_results.push_back(run_test("Star graph (20)", g, 0));
    }
    
    {
        Graph g = create_complete_graph(15);
        test_results.push_back(run_test("Complete graph (15)", g, 0));
    }
    
    // Medium graphs
    {
        Graph g = create_grid_graph(30);  // 900 vertices
        test_results.push_back(run_test("Grid 30x30 (900)", g, 0));
    }
    
    {
        Graph g = create_line_graph(1000);
        test_results.push_back(run_test("Line graph (1000)", g, 0));
    }
    
    {
        Graph g = create_star_graph(5000);
        test_results.push_back(run_test("Star graph (5000)", g, 0));
    }
    
    // Large graphs
    {
        Graph g = create_grid_graph(100);  // 10,000 vertices
        test_results.push_back(run_test("Grid 100x100 (10k)", g, 5050));  // center
    }
    
    {
        Graph g = create_grid_graph(200);  // 40,000 vertices
        test_results.push_back(run_test("Grid 200x200 (40k)", g, 20100));  // center
    }
    
    // Print results
    int passed = 0;
    int failed = 0;
    
    for (const auto& test : test_results) {
        bool pass = test.correctness_pass && test.determinism_pass;
        std::cout << "[" << (pass ? "PASS" : "FAIL") << "] " << test.name 
                  << " (n=" << test.size << ")" << std::endl;
        
        if (!pass) {
            std::cout << "  Error: " << test.error_msg << std::endl;
            failed++;
        } else {
            std::cout << "  Correctness: ✓  Determinism: ✓" << std::endl;
            if (!test.hashes.empty()) {
                std::cout << "  Hash: " << hash_to_hex(test.hashes[0]) << std::endl;
            }
            passed++;
        }
        std::cout << std::endl;
    }
    
    std::cout << "=== Summary ===" << std::endl;
    std::cout << "Passed: " << passed << "/" << (passed + failed) << std::endl;
    std::cout << "Failed: " << failed << "/" << (passed + failed) << std::endl;
    
    // Write summary to file
    std::ofstream summary("run_summary.txt");
    summary << "BFS Differential Test Results\n";
    summary << "==============================\n\n";
    
    for (const auto& test : test_results) {
        bool pass = test.correctness_pass && test.determinism_pass;
        summary << "[" << (pass ? "PASS" : "FAIL") << "] " << test.name 
                << " (n=" << test.size << ")\n";
        
        if (!pass) {
            summary << "  Error: " << test.error_msg << "\n";
        } else {
            summary << "  Correctness: PASS\n";
            summary << "  Determinism: PASS (3 runs)\n";
            if (!test.hashes.empty()) {
                summary << "  Hashes:\n";
                for (size_t i = 0; i < test.hashes.size(); i++) {
                    summary << "    Run " << i << ": " << hash_to_hex(test.hashes[i]) << "\n";
                }
            }
        }
        summary << "\n";
    }
    
    summary << "Summary: " << passed << " passed, " << failed << " failed\n";
    summary.close();
    
    // Performance tests
    if (run_perf) {
        std::cout << "\n=== Performance Benchmarks ===" << std::endl;
        std::vector<PerfResult> perf_results;
        
        {
            Graph g = create_grid_graph(200);
            perf_results.push_back(run_perf_test("Grid 200x200", g, 20100));
        }
        
        {
            Graph g = create_grid_graph(300);
            perf_results.push_back(run_perf_test("Grid 300x300", g, 45150));
        }
        
        std::ofstream perf_file("perf.txt");
        perf_file << "BFS Performance Results\n";
        perf_file << "=======================\n\n";
        
        for (const auto& perf : perf_results) {
            std::cout << perf.name << " (n=" << perf.size << ")" << std::endl;
            std::cout << "  Sequential: " << std::fixed << std::setprecision(2) 
                     << perf.seq_time_ms << " ms" << std::endl;
            std::cout << "  Parallel (" << perf.num_threads << " threads): " 
                     << perf.par_time_ms << " ms" << std::endl;
            std::cout << "  Speedup: " << std::setprecision(2) << perf.speedup << "x" << std::endl;
            std::cout << "  Efficiency: " << std::setprecision(1) 
                     << (perf.speedup / perf.num_threads * 100) << "%" << std::endl;
            std::cout << std::endl;
            
            perf_file << perf.name << " (n=" << perf.size << ")\n";
            perf_file << "  Sequential: " << std::fixed << std::setprecision(2) 
                     << perf.seq_time_ms << " ms\n";
            perf_file << "  Parallel (" << perf.num_threads << " threads): " 
                     << perf.par_time_ms << " ms\n";
            perf_file << "  Speedup: " << std::setprecision(2) << perf.speedup << "x\n";
            perf_file << "  Efficiency: " << std::setprecision(1) 
                     << (perf.speedup / perf.num_threads * 100) << "%\n\n";
        }
        
        perf_file.close();
    }
    
    return (failed > 0) ? 1 : 0;
}
